src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlBulkCopy.cs

```diff
@@ -965,37 +985,66 @@ private object GetValueFromSourceRow(int destRowIndex, out bool isSqlType, out b
                                 Debug.Assert(!(value is INullable) || !((INullable)value).IsNull, "IsDBNull returned false, but GetValue returned a null INullable");
                             }
 #endif
-                            return value;
+                            return WriteValueAsync(value, destRowIndex, isSqlType, isDataFeed, isNull);
                         }
                     }
                     else
                     {
-                        isDataFeed = false;
-
                         IDataReader rowSourceAsIDataReader = (IDataReader)_rowSource;
-
-                        // Only use IsDbNull when streaming is enabled and only for non-SqlDataReader
-                        if ((_enableStreaming) && (_sqlDataReaderRowSource == null) && (rowSourceAsIDataReader.IsDBNull(sourceOrdinal)))
+                        // previously, IsDbNull was only invoked in a non-streaming scenario with a non-SqlDataReader.
```

Comment:
@cheenamalhotra @Wraith2 starting a conversation here to keep this tied together -- I will create a benchmark around this logic that inserts a million records in a loop until i hit about the minute mark.

I will do this in a couple of other branches in my repo and will post some links to the code here -- should be able to get this done today.

Reply:
If IsDbNull proves to be problematic, there could be an optimization around only invoking it for fields that are possibly null (e.g. if `GetFieldType` returns `typeof(int)` we don't need to check

Reply:
If you're inspecting the SchemaTable there is IsNullable field, if it's not present you should be able to skip the dbnull checks.